README.md
[![Build Status](https://travis-ci.org/jzaefferer/node-testswarm.svg?branch=master)](https://travis-ci.org/jzaefferer/node-testswarm) [![NPM version](https://badge.fury.io/js/testswarm.svg)](http://badge.fury.io/js/testswarm)

# node-testswarm

Nodejs module for interacting with TestSwarm

## Getting Started
Install the module with:

	npm install --save-dev testswarm

```javascript
var testswarm = require( "./lib/testswarm" ),
	testUrl = "http://localhost/jquery-core/test/",
	runs = {};

["attributes", "callbacks"].forEach(function (suite) {
	runs[suite] = testUrl + "?module=" + suite;
});

testswarm.createClient({
	url: "http://localhost/testswarm/"
})
.addReporter( testswarm.reporters.cli )
.auth({
	id: "example",
	token: "yourauthtoken"
})
.addjob(
	{
		name: "node-testswarm test job",
		runs: runs,
		browserSets: [ "example" ],
	}, function( err, passed ) {
		if ( err ) {
			throw err;
		}
		process.exit( passed ? 0 : 1 );
	}
);
```

For local testing, copy `sample-test.js` to `test.js` and modify to match your local TestSwarm setup.

## API

### createClient({ url })

* `Object config`
 - `String url` - Url to root of TestSwarm install.

### Client#addReporter( reporter )

* `Object reporter` - usually `testswarm.reporters.cli`, unless you want to use a custom reporter

### Client#auth({ id, token })

* `Object auth`
 - `String id` - Username of TestSwarm account.
 - `String token` - Authentication token of account.

### Client#addjob( options, callback )

* `Object options`
 - `String name` - name of this job
 - `Number runMax` - [optional] how often failed tests should rerun
 - `Object runs` -  Run urls by run name.
 - `Array|String browserSets` - which sets to test against
 - `Number pollInterval` -  [optional] In milliseconds, default 5 seconds.
 - `Number timeout` -  [optional] In milliseconds, default 15 minutes.
* `Function( Object err, Boolean passed, Object results ) callback`

## License
Copyright (c) 2014 Jörn Zaefferer
Licensed under the MIT license.


History.md
# Release History

## 1.1.2 / 2020-06-10

* Fix invalid addjob() requests that regressed in v1.1.1. (Michał Gołębiowski-Owczarek)

## 1.1.1 / 2020-06-06

* Change 'request' dependency from 2.12.0 to 2.88.0. (Michał Gołębiowski-Owczarek)
* Write API docs in the README. (Jörn Zaefferer)

## 1.1.0 / 2013-05-14

* Change authentication request to use `authID` instead of `authUsername`
  to accomodate changes in upstream TestSwarm 1.0.0-alpha. (Timo Tijhof)

## 1.0.3 / 2013-04-30

* Fix the timeout error to display the timeout instead of undefined. (Daniel Herman)

## 1.0.1 / 2013-03-17

* (issue #9) Always invoke the addjob callback with a boolean "passed" value. (Timo Tijhof)

## 1.0.0 / 2012-12-16

* Complete rewrite providing a new API, not backwards compatible. (Timo Tijhof)

## 0.3.0 / 2012-11-29

* Refactor logic, backwards compatible. (Jörn Zaefferer)
* Clean up output.

## 0.2.3 / 2012-06-16

* Add try-catch around JSON.parse call, in case server returns invalid JSON.

## 0.2.2 / 2012-05-02

* Simplify results output (again).
* Don't care about trailing slash in URL.

## 0.2.1 / 2012-05-01

* browserSets parameter is now an array.
* Made timed out runs result in failure.
* Support for all failure types othan than "failed" ("error", "timedout", ..).
* Simplify results output.

## 0.2.0 / 2012-04-30

* Global timeout.
* Check all uaRuns.
* Proper URLs back to TestSwarm.
* Some additional error handling.

## 0.1.1 / 2012-xx-xx

* Fix engine property in package.json.

## 0.1.0 / 2012-xx-xx

* Initial release.


test.js
var myTestPage = "http://localhost/jquery-core-checkout/test/",
	mySuites = ["attributes", "callbacks"];
require( "./testswarm" )( {
	url: "http://localhost/testswarm/",
	pollInterval: 1000,
	// 1 minute
	timeout: 1000 * 60 * 1,
	done: function( passed ) {
		process.exit( passed ? 0 : 1 );
	}
}, {
	authUsername: "swarmuser",
	authToken: "yourauthtoken",
	jobName: "node-testswarm test job",
	runMax: 2,
	"runNames[]": mySuites,
	"runUrls[]": mySuites.map(function (suite) {
		return myTestPage + "?module=" + suite;
	}),
	"browserSets[]": ["example"]
});


lib/reporters.js
module.exports = {
	cli: function( testswarm ) {
		testswarm.on( "log", console.log.bind( console ) );

		//testswarm.on( "verbose", console.log.bind( console ) );

		testswarm.on( "addjob-poll", function() {
			process.stdout.write( "." );
		} );

		testswarm.on( "addjob-ready", function( err, passed, results ) {
			var name, result, uaID;
			// Line break to return from the dot-writes
			process.stdout.write( "\n" );

			console.log( "State: " + ( err || "Finished" ));
			if ( results ) {
				console.log( "Results: " );
				for ( name in results ) {
					console.log( "\t* " + name );
					result = results[name];
					for ( uaID in result ) {
						console.log( "\t  " + uaID + ": " + result[uaID] );
					}
				}
			}
		} );
	},

  junit: function (testswarm) {
    // same as the CLI reporter, but also writes a file TEST-javascript.xml in the junit format

    testswarm.on("log", console.log.bind(console));

    //testswarm.on( "verbose", console.log.bind( console ) );

    testswarm.on("addjob-poll", function () {
      process.stdout.write(".");
    });

    testswarm.on("addjob-ready", function (err, passed, results) {
      var name, result, uaID;
      var count_all = 0, count_fail = 0;
      var body = '';

      // Line break to return from the dot-writes
      process.stdout.write("\n");

      console.log("State: " + ( err || "Finished" ));
      if (results) {
        console.log("Results: ");
        for (name in results) {
          console.log("\t* " + name);
          result = results[name];
          for (uaID in result) {
            console.log("\t  " + uaID + ": " + result[uaID]);

            // JUNIT!!
            count_all += 1;
            body += '<testcase classname="javascript.TestSwarm" name="' + name + ' - '
              + uaID + '" time="0">';
            if (result[uaID].indexOf('passed') != 0) {
              count_fail += 1;
              body += '<failure />';
            }
            body += '<system-out>' + result[uaID] + '</system-out>' + '</testcase>\n';
          }
        }
      }
      var report = '<?xml version="1.0" ?>\n';
      report += '<testsuite errors="0" failures="' + count_fail + '" name="javascript.TestSwarm" tests="'
        + count_all + '" time="0">\n';
      report += body;
      report += '</testsuite>';

      // write out the xml
      var fs = require("fs");
      fs.writeFileSync('TEST-javascript.xml', report, 'utf-8');
    });
  }
};


lib/testswarm.js
var request = require( "request" ),
	querystring = require( "querystring" ).stringify,
	url = require( "url" ),
	util = require( "util" ),
	events = require( "events" );

function extend( a, b ) {
	for ( var key in b ) {
		a[key] = b[key];
	}
	return a;
}

/**
 * Utility for polling the API.
 * @param {Object} params
 * - {number} interval
 * - {number} timeout
 * - {Function} makeRequest
 * - {Function} doContinue
 * - {Function} onSuccess [optional] Called when continue returns false.
 * - {Function} onTimeout [optional] Called if continue returns true after timeout is reached.
 */
function pollJSON( params ) {
	var tsStart = Date.now();

	params = extend({
		onSuccess: function() {},
		onTimeout: function() {}
	}, params );

	function handleRequest( err, resp, body ) {
		var data;
		if ( err ) {
			throw err;
		}
		try {
			data = JSON.parse( body );
		} catch ( e ) {
			throw e;
		}
		if ( params.doContinue( data ) ) {
			if ( tsStart + params.timeout < Date.now() ) {
				params.onTimeout( data );
			} else {
				setTimeout( function() {
					params.makeRequest( handleRequest );
				}, params.interval );
			}
		} else {
			params.onSuccess( data );
		}
	}

	params.makeRequest( handleRequest );
}

function isJobDone( job ) {
	return !job.runs.filter(function( run ) {
		return Object.keys(run.uaRuns).filter(function( uaRun ) {
			return (/new|progress/).test( run.uaRuns[ uaRun ].runStatus );
		}).length;
	}).length;
}

/**
 * @param {TestSwarm} testswarm
 * @param {Object} options From TestSwarm..addjob.
 * @param {Object} jobInfo
 * @param {Function( {mixed} err, {Object} job )} callback
 */
function pollJobStatus( testswarm, options, jobInfo, callback ) {
	var config = testswarm.config,
		i = 1;
	pollJSON({
		interval: options.pollInterval,
		timeout: options.timeout,
		makeRequest: function( handleRequest ) {
			request.get( config.url + "/api.php?" + querystring({
				action: "job",
				item: jobInfo.id
			}), handleRequest);
		},
		doContinue: function( data ) {
			if ( !data.job ) {
				throw new Error( "Invalid API response, can't continue. Response was: " + data );
			}
			testswarm.emit( "addjob-poll", ++i );
			return !isJobDone( data.job );
		},
		onTimeout: function( data ) {
			callback( "Timed out after " + options.timeout + "ms", data.job );
		},
		onSuccess: function( data ) {
			callback( null, data.job );
		}
	});
}

/**
 * Get url to root of testswarm server (without the path to the installation, if any).
 *
 * @param {TestSwarm} testswarm
 * @return {string} Domain root of url.
 */
function getServerRoot( testswarm ) {
	var obj = url.parse( testswarm.config.url );
	obj.href = obj.pathname = obj.search = obj.path = obj.query = obj.hash = undefined;
	// Normalise url to no trailing slash
	return url.format( obj ).replace( /\/$/, "" );
}

/**
 * Add authentication parameters to request data.
 *
 * @param {TestSwarm} testswarm
 * @param {Object} data
 * @return {Object} data
 */
function authenticateRequest( testswarm, data ) {
	if ( !testswarm.auth ) {
		throw new Error( "Authentication required, call .auth() first." );
	}
	data.authID = testswarm.auth.id;
	data.authToken = testswarm.auth.token;
	return data;
}

function keysMissingFrom( obj, keys ) {
	return keys.filter(function( val ) {
		if ( obj[ val ] === undefined ) {
			return true;
		}
		return false;
	});
}

/**
 * @constructor
 * @extends EventEmitter
 * @param {Object} config
 * - {string} url Url to root of TestSwarm install.
 * @emits {string} log Regular output regarding progress of performed actions.
 * @emits {string} verbose Verbose output for debugging.
 */
function TestSwarm( config ) {
	var missing = keysMissingFrom( config, [ "url" ] );
	if ( missing.length ) {
		throw new Error( "Invalid configuration, missing key '" + missing[0] + "'." );
	}

	// Normalise url to no trailing slash
	config.url = url.format( url.parse( config.url ) ).replace( /\/$/, "" );
	this.config = config;
}

util.inherits( TestSwarm, events.EventEmitter );

TestSwarm.prototype.addReporter = function( reporter ) {
	reporter( this );
	return this;
};

/**
 * @param {Object} auth
 * - {string} id Username of TestSwarm account.
 * - {string} token Authentication token of account.
 */
TestSwarm.prototype.auth = function( auth ) {
	var missing = keysMissingFrom( auth, [ "id", "token" ] );
	if ( missing.length ) {
		throw new Error( "Invalid authentication, missing key '" + missing[0] + "'." );
	}

	this.auth = auth;
	return this;
};

/**
 * @method
 * @param {Object} options
 * - {string} name
 * - {number} runMax [optional]
 * - {Object} runs Run urls by run name.
 * - {Array|string} browserSets
 * - {number} pollInterval [optional] In milliseconds.
 * - {number} timeout [optional] In milliseconds.
 * @param {Function( {mixed} err, {boolean} passed, {Object} results )} callback
 * @emits addjob-created ( {Object} jobInfo )
 * @emits addjob-poll ( {number}  iteration ) Starts with 1.
 * @emits addjob-ready ( {boolean} passsed, {Object} results )
 */
TestSwarm.prototype.addjob = function( options, callback ) {
	var testswarm, missing, runNames, runUrls, browserSets;

	testswarm = this;
	missing = keysMissingFrom( options, [ "name", "runs", "browserSets" ] );
	if ( missing.length ) {
		callback( "Invalid options, missing key '" + missing[0] + "'." );
		return;
	}

	options = extend({
		runMax: 2,
		pollInterval: 5000, // 5 seconds
		timeout: 15 * 60 * 1000 // 15 minutes
	}, options);

	runNames = Object.keys( options.runs );
	runUrls = runNames.slice().map(function( name ) {
		return options.runs[name];
	});

	browserSets = Array.isArray( options.browserSets ) ?
		options.browserSets :
		[ options.browserSets ];

	request.post({
		url: testswarm.config.url + "/api.php",
		form: authenticateRequest( testswarm, {
			action: "addjob",
			jobName: options.name,
			runMax: options.runMax,
			runNames: runNames,
			runUrls: runUrls,
			browserSets: browserSets
		})
	}, function( err, resp, body ) {
		var data, jobInfo;
		if ( err ) {
			callback( err, false );
			return;
		}
		try {
			data = JSON.parse( body );
		} catch ( e ) {
			callback( "Failed parsing body as json, was: " + body, false );
		}
		if ( !data.addjob ) {
			callback( "API returned error, can't continue. Response was: " + body, false );
			return;
		}
		jobInfo = data.addjob;
		testswarm.emit( "log", "Submitted job " + jobInfo.id + " " + testswarm.config.url + "/job/" + jobInfo.id );
		testswarm.emit( "log", "Polling for results" );
		testswarm.emit( "verbose", JSON.stringify( jobInfo ) );
		testswarm.emit( "addjob-created", jobInfo );
		pollJobStatus( testswarm, options, jobInfo, function( err, job ) {
			var passed = true, results = {};

			function runStats( uaID, uaRun ) {
				var val = uaRun.runStatus;
				if ( uaRun.runResultsUrl ) {
					val +=  " (" + uaRun.runResultsLabel + "). " + getServerRoot( testswarm ) + uaRun.runResultsUrl;
				}
				return val;
			}

			job.runs.filter(function( run ) {
				results[run.info.name] = {};
				for ( var uaID in run.uaRuns ) {
					results[run.info.name][uaID] = runStats( uaID, run.uaRuns[uaID] );
					if ( passed && run.uaRuns[uaID].runStatus !== "passed" ) {
						// "new", "failed", "error", "timedout", ..
						passed = false;
					}
				}
			});

			testswarm.emit( "addjob-ready", err, passed, results );
			callback( err, passed, results );
		} );
	});
};

module.exports = {
	createClient: function( config ) {
		return new TestSwarm( config );
	},
	reporters: require( "./reporters" )
};
